app/src/main/java/org/hitlabnz/sensor_fusion_demo/Compass3D.java

```diff
@@ -0,0 +1,310 @@
+package org.hitlabnz.sensor_fusion_demo;
+
+import static android.content.Context.SENSOR_SERVICE;
+import android.hardware.SensorManager;
+import com.badlogic.gdx.ApplicationAdapter;
+import com.badlogic.gdx.Gdx;
+import com.badlogic.gdx.Input;
+import com.badlogic.gdx.InputAdapter;
+import com.badlogic.gdx.Version;
+import com.badlogic.gdx.graphics.Color;
+import com.badlogic.gdx.graphics.GL20;
+import com.badlogic.gdx.graphics.PerspectiveCamera;
+import com.badlogic.gdx.graphics.Texture;
+import com.badlogic.gdx.graphics.VertexAttributes.Usage;
+import com.badlogic.gdx.graphics.g2d.BitmapFont;
+import com.badlogic.gdx.graphics.g2d.SpriteBatch;
+import com.badlogic.gdx.graphics.g3d.Environment;
+import com.badlogic.gdx.graphics.g3d.Material;
+import com.badlogic.gdx.graphics.g3d.Model;
+import com.badlogic.gdx.graphics.g3d.ModelBatch;
+import com.badlogic.gdx.graphics.g3d.ModelInstance;
+import com.badlogic.gdx.graphics.g3d.attributes.ColorAttribute;
+import com.badlogic.gdx.graphics.g3d.attributes.TextureAttribute;
+import com.badlogic.gdx.graphics.g3d.environment.PointLight;
+import com.badlogic.gdx.graphics.g3d.utils.ModelBuilder;
+import com.badlogic.gdx.math.MathUtils;
+import com.badlogic.gdx.math.Quaternion;
+import com.badlogic.gdx.math.Vector3;
+import com.badlogic.gdx.scenes.scene2d.Stage;
+import com.badlogic.gdx.scenes.scene2d.ui.Label;
+import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle;
+import com.badlogic.gdx.graphics.g3d.model.Node;
+import com.badlogic.gdx.scenes.scene2d.ui.Table;
+import com.badlogic.gdx.utils.ScreenUtils;
+import org.hitlabnz.sensorfusionlib.orientationProvider.*;
+import org.hitlabnz.sensorfusionlib.representation.*;
+
+public class Compass3D extends ApplicationAdapter {
+    private final String TAG = Compass3D.class.getName();
+    private final SensorSelectionActivity activity;
+    private final SensorManager sm;
+    private Environment lights;
+    private PerspectiveCamera cam;
+    private ModelBatch modelBatch;
+    private Model model;
+    private ModelInstance instance;
+    private Texture north, east, south, west, zenith, nadir;
+    private SpriteBatch batch;
+    private BitmapFont font;
+    private Stage stage;
+    private Label label;
+    private MatrixF4x4 rm;
+    private float[] RotMat, ModMat, angles;
+    private Table table;
+    private String currentOrientationProvider;
+
+    // we have 5 different orientation providers suitable for 3D compass,
+    // Calibrated Gyroscope cannot be used for a compass application
+    private final int nProviders = 5;
```

Comment:
I am well aware, that a gyroscope cannot be used for a compass application. But for educational purposes, I explicitly provided it here too, that everyone can see and test, that just a gyroscope will give you some uninitialized relative position only. So please include it again.

Reply:
> I am well aware, that a gyroscope cannot be used for a compass application. But for educational purposes, I explicitly provided it here too, that everyone can see and test, that just a gyroscope will give you some uninitialized relative position only. So please include it again.

Yes, we can add gyroscope back to the "game" but this will require additional analisys, since using the current code it makes the 3D compass to appear totally incorrent.

Reply:
> ...since using the current code it makes the 3D compass to appear totally incorrect.

I'm aware of this, and it's OK. It was like this before and I wanted it like this. I think it is a great way to demonstrate, that the gyroscope alone will not give you a compass.